File: chzzk.js
/**
 * @fileoverview 치지직 API 모듈
 * 팔로워 정보 및 프로필 데이터를 가져옵니다.
 * @module chzzk
 */

const axios = require('axios');
const config = require('./config');
const auth = require('./auth');
const logger = require('./logger');

/**
 * 현재 로그인된 사용자의 프로필 ID (캐시)
 * @type {string}
 */
let profileId = '';

/**
 * 사용자 프로필 ID를 가져옵니다.
 * @async
 * @returns {Promise<string>} 사용자 프로필 ID (userIdHash)
 * @throws {Error} 인증 실패 또는 네트워크 오류 시
 */
async function getProfileId() {
    if (profileId) return profileId;

    try {
        const apiUrl = `${config.api.naverGame}/nng_main/v1/user/getUserStatus`;
        const cookies = await auth.getAuthCookies();

        const response = await axios.get(apiUrl, {
            headers: {
                Cookie: Object.entries(cookies).map(([name, value]) => `${name}=${value}`).join('; '),
            },
            timeout: 5000
        });

        logger.info('[Chzzk] 프로필 데이터 수신:', response.data);
        profileId = response.data.content.userIdHash;
        return profileId;
    } catch (error) {
        logger.error('[Chzzk] 프로필 ID 가져오기 실패:', error.message);
        if (error.response && (error.response.status === 401 || error.response.status === 403)) {
            logger.error('[Chzzk] 세션 만료됨');
            auth.clearSessionData();
            profileId = '';
        }
        throw error;
    }
}

/**
 * 팔로워 목록을 가져옵니다.
 * @async
 * @param {number} [page=0] - 페이지 번호 (0부터 시작)
 * @param {number} [size=10] - 페이지당 팔로워 수
 * @returns {Promise<{code: number, content: {data: Array}}>} 팔로워 응답 데이터
 * @throws {Error} 인증 실패 또는 네트워크 오류 시
 */
async function getFollowers(page = 0, size = 10) {
    try {
        const currentProfileId = await getProfileId();
        const apiUrl = `${config.api.chzzk}/manage/v1/channels/${currentProfileId}/followers?page=${page}&size=${size}&userNickname=`;

        const cookies = await auth.getAuthCookies();

        const response = await axios.get(apiUrl, {
            headers: {
                Cookie: Object.entries(cookies).map(([name, value]) => `${name}=${value}`).join('; '),
            },
            timeout: 5000
        });

        return response.data;
    } catch (error) {
        logger.error('[Chzzk] 팔로워 가져오기 실패:', error.message);
        if (error.response && (error.response.status === 401 || error.response.status === 403)) {
            logger.error('[Chzzk] 세션 만료됨');
            auth.clearSessionData();
            profileId = '';
        }
        throw error;
    }
}

/**
 * 캐시된 프로필 ID를 초기화합니다.
 * @returns {void}
 */
function resetProfileId() {
    profileId = '';
}

module.exports = {
    getProfileId,
    getFollowers,
    resetProfileId
};

File: src-tauri/src/chzzk.rs
use reqwest::header::{HeaderMap, HeaderValue, COOKIE, USER_AGENT};
use serde::{Deserialize, Serialize};
use crate::state::CookieData;

#[derive(Debug, Deserialize)]
struct UserStatusResponse {
    code: i32,
    message: Option<String>,
    content: Option<UserStatusContent>,
}

#[derive(Debug, Deserialize)]
struct UserStatusContent {
    #[serde(rename = "userIdHash")]
    user_id_hash: String,
    nickname: String,
}

pub async fn get_profile_id(client: &reqwest::Client, cookies: &CookieData) -> Result<(String, String), String> {
    let url = "https://comm-api.game.naver.com/nng_main/v1/user/getUserStatus";
    
    let cookie_str = format!("NID_AUT={}; NID_SES={}", cookies.nid_aut, cookies.nid_ses);
    
    let mut headers = HeaderMap::new();
    headers.insert(COOKIE, HeaderValue::from_str(&cookie_str).map_err(|e| e.to_string())?);
    headers.insert(USER_AGENT, HeaderValue::from_static("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"));

    // let client = reqwest::Client::new(); // Use passed client
    let res = client.get(url)
        .headers(headers)
        .send()
        .await
        .map_err(|e| e.to_string())?;
        
    if !res.status().is_success() {
        return Err(format!("API Error: {}", res.status()));
    }
    
    let body: UserStatusResponse = res.json().await.map_err(|e| e.to_string())?;
    
    if body.code == 200 {
        if let Some(content) = body.content {
            return Ok((content.user_id_hash, content.nickname));
        }
    }
    
    Err(format!("Auth Failed: {:?}", body.message))
}

#[derive(Debug, Deserialize, Serialize, Clone, Default)]
#[serde(default)]
pub struct FollowerResponse {
    pub code: i32,
    pub message: Option<String>,
    pub content: Option<FollowerContent>,
}

#[derive(Debug, Deserialize, Serialize, Clone, Default)]
#[serde(default)]
pub struct FollowerContent {
    pub data: Vec<FollowerItem>,
}

#[derive(Debug, Deserialize, Serialize, Clone, Default)]
#[serde(default)]
pub struct FollowerItem {
    pub user: User,
    #[serde(rename = "followingSince")]
    pub following_since: String,
}

#[derive(Debug, Deserialize, Serialize, Clone, Default)]
#[serde(default)]
pub struct User {
    #[serde(rename = "userIdHash")]
    pub user_id_hash: String,
    pub nickname: String,
    #[serde(rename = "profileImageUrl")]
    pub profile_image_url: Option<String>,
}

pub async fn get_followers(client: &reqwest::Client, cookies: &CookieData, user_id_hash: &str) -> Result<FollowerResponse, String> {
    let url = format!("https://api.chzzk.naver.com/manage/v1/channels/{}/followers?page=0&size=10&userNickname=", user_id_hash);
    
    let cookie_str = format!("NID_AUT={}; NID_SES={}", cookies.nid_aut, cookies.nid_ses);
    
    let mut headers = HeaderMap::new();
    headers.insert(COOKIE, HeaderValue::from_str(&cookie_str).map_err(|e| e.to_string())?);
    headers.insert(USER_AGENT, HeaderValue::from_static("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"));

    // let client = reqwest::Client::new();
    let res = client.get(url)
        .headers(headers)
        .send()
        .await
        .map_err(|e| e.to_string())?;
        
    if !res.status().is_success() {
        return Err(format!("API Error: {}", res.status()));
    }
    
    let body_text = res.text().await.map_err(|e| e.to_string())?;
    println!("[Chzzk API] Raw Response: {}", &body_text.chars().take(2000).collect::<String>());
    
    let body: FollowerResponse = serde_json::from_str(&body_text).map_err(|e| {
        eprintln!("[Chzzk API] Parse Error: {}", e);
        e.to_string()
    })?;
    Ok(body)
}
